Grenades implementation

Badulundabad opened the pull request.

Badulundabad committed:
Added GrenadeMissionLogic and GrenadeScript

Added the grenade usage set;
Created the GrenadeMissionLogic and GrenadeScript.

Badulundabad committed:
Modified the grenade damege formula

Badulundabad committed:
Fixed grenade bouncing back

Badulundabad committed:
Fixed grenade bouncing back

martkartasev commented on CSharpSourceCode/Battle/Grenades/GrenadesMissionLogic.cs:
```diff
@@ -0,0 +1,63 @@
+            base.OnAgentHit(affectedAgent, affectorAgent, damage, affectorWeapon);
+            if (affectorWeapon.Item != null && affectorWeapon.Item.Name.Contains("Grenade"))
+            {
+                Mission.Missile grenade = Mission.Current.Missiles.ToList().First(m => m.Weapon.Item.Name.Contains("Grenade"));
```
This will break when there is more than one grenade in the mission.

Something like this should assure you find the correct missile not just the first one that you happen to iterate to.

> var id = affectorWeapon.Item.Id;
> Mission.Missile grenade = Mission.Current.Missiles.ToList().First(m => m.Weapon.Item.Id == id);

Im also not sure about potential performance impact of running this on every hit? 
Thoughts @hunharibo @tristanka @Z3rca ?

Badulundabad replied:
Okay, i'll test with id. But about using OnAgentHit...i'm not sure that there is another way. At least i'll try to handle missile collision reaction.

martkartasev commented on CSharpSourceCode/Battle/Grenades/GrenadesMissionLogic.cs:
```diff
@@ -0,0 +1,63 @@
+        public override void OnMissionTick(float dt)
+        {
+            base.OnMissionTick(dt);
+            if (Input.IsKeyPressed(InputKey.L))
```
Should be removed before final merge.

Badulundabad replied:
Sure. I want to move this to DebugMissionLogic i am working on.

martkartasev reviewed (changes requested):
In general I think the structure is good as it follows the rest of the code structure.

We might be able to cause the damage effect with some of the new stuff from @hunharibo AbilityRework.

What do you guys think is best. Just leave it for now and figure how to merge the two later when both are in development?

@Z3rca @hunharibo @tristanka

martkartasev commented on CSharpSourceCode/Battle/TOWBattleUtilities.cs:
```diff
@@ -34,7 +32,24 @@ public static void DamageAgentsInArea(Vec2 center, float radius, int minDamage,
             }
         }
-
+        public static void DamageAgentsWithGrenade(Vec2 center, float radius, int maxDamage = 0, Agent damager = null, bool enemyOnly = true)
```
Is there a way to replicate this in the new ability system @hunharibo ?

Badulundabad committed:
Made some changes. Reade the description.

Fixed the issue when a grenade doesn't bounce back from a shield;
Added Weapon.Item.Id changing. Now we able to use more grenades, but with only one grenade for each MissionWeapon;
Added "Tick" sound for grenade.

Badulundabad committed:
Fixed the grenade logics, reworked by TriggeredEffect

hunharibo reviewed (approved):
Lets get this thing merged!

Z3rca commented on CSharpSourceCode/Battle/TriggeredEffect/Scripts/GrenadeScript.cs:
```diff
@@ -0,0 +1,65 @@
+                hasLaunched = true;
+                _tickSound.Play();
+            }
+            if (hasLaunched && explosionTimer % 40 == 0)
```
magic numbers

Z3rca reviewed (approved):
I dont know if Throw is better, or more confusing for people who are more used to the System naming. We have some magic numbers all over the code... maybe we should make a setting XML which get set on start. Just a thought, don't need to act here for now

Badulundabad committed:
Added knocking down

Badulundabad committed:
Rebase from development
